src/main/java/bio/terra/common/ValidationUtils.java

```diff
@@ -80,6 +81,30 @@ public static String requireNotBlank(String value, String errorMsg) {
     return value;
   }
 
+  public static LinkedHashMap<String, String> validateMatchingColumnDataTypes(
+      RelationshipTermModel fromTerm, RelationshipTermModel toTerm, List<TableModel> tables) {
+    LinkedHashMap<String, String> termErrors = new LinkedHashMap<>();
+    Optional<ColumnModel> fromColumn = retrieveColumnModelFromTerm(fromTerm, tables);
+    Optional<ColumnModel> toColumn = retrieveColumnModelFromTerm(toTerm, tables);
+    if (fromColumn.isPresent() && toColumn.isPresent()) {
+      TableDataType fromColumnDataType = fromColumn.get().getDatatype();
+      TableDataType toColumnDataType = toColumn.get().getDatatype();
+      if (!fromColumnDataType.equals(toColumnDataType)) {
+        termErrors.put(
+            "RelationshipDatatypeMismatch",
+            String.format(
+                "Column data types in relationship must match: Column %s.%s has data type %s and Column %s.%s has data type %s",
+                fromTerm.getTable(),
+                fromTerm.getColumn(),
+                fromColumnDataType,
+                toTerm.getTable(),
+                toTerm.getColumn(),
+                toColumnDataType));
+      }
+    }
```

Comment:
`Optional.get()` can be avoided here too.
```suggestion
    retrieveColumnModelFromTerm(fromTerm, tables).ifPresent(
        fromColumn ->
            retrieveColumnModelFromTerm(toTerm, tables).ifPresent(toColumn -> {
                  TableDataType fromColumnDataType = fromColumn.getDatatype();
                  TableDataType toColumnDataType = toColumn.getDatatype();
                  if (!fromColumnDataType.equals(toColumnDataType)) {
                    termErrors.put(
                        "RelationshipDatatypeMismatch",
                        String.format(
                            "Column data types in relationship must match: Column %s.%s has data type %s and Column %s.%s has data type %s",
                            fromTerm.getTable(),
                            fromTerm.getColumn(),
                            fromColumnDataType,
                            toTerm.getTable(),
                            toTerm.getColumn(),
                            toColumnDataType));
                  }
                }));
```